Add options page

jeongm-in opened the pull request:
Pretty straightforward PR.

It has been a painful past few days, but I am glad it's over now. 

Options page looks like below image:
<img width="932" alt="option" src="https://user-images.githubusercontent.com/8725873/60867671-08e61f00-a1f1-11e9-8835-c5946c998b7c.PNG">

Background color is set to reflect our team's taste. 
The toggle is grabbed from [react-toggle](https://github.com/aaronshaf/react-toggle).
`OptionToggle.tsx` contains component for option toggles, which can be used for any `boolean` typed setting fields in `chrome.storage.sync`. Should we need more toggles, we can easily add any number of toggles.

On changing the toggle, `onChange` function will fire and set the value in `chrome.storage.sync` as well. The value of toggle is stored in `state` of the component as well, and this value is loaded from `chrome.storage.sync` on `componentDidMount()` lifecycle. (It was good to finally understand different lifecycles of components.)

While writing this PR, I just realized we need to add another card to tell users who we are. 
Also a bit more tweak to reduce the gap between `Logo` and the first `Card` might be better.

Special thanks to @Bartleby2718  for helping me through the coding blockade.

jeongm-in committed:
Add skeleton html and css for the static options page close #65

jeongm-in committed:
commit for share

jeongm-in committed:
configure webpack to render react in options.html close #64

jeongm-in committed:
Create new promo and icon close #72

jeongm-in committed:
Add new components for the options page close#71

jeongm-in committed:
Support loading third party css files from node_modules close #74

jeongm-in committed:
Add more fields in locales for options page close #73

jeongm-in committed:
Update links to icon in html and manifest close #76

jeongm-in committed:
Add more fields in locales for About card close #84

jeongm-in committed:
Nest components for better modularity close #83

jeongm-in committed:
Tweak color scheme and responsiveness of options page #79

jeongm-in committed:
Add Github icon to use as button close #85

jeongm-in commented:
PR addendum
1. Now components are truly isolated for better modular behavior. 
  - See [a4dbe3a](https://github.com/jeongm-in/webmark/pull/77/commits/a4dbe3ae24f2e44cef2eb492efad4626f05c5ed2)
  - `{this.props.children` allows nested react components to render. 
  - Should we need more icons for the `homepage` field, `LinkFIeld` component can be broken into `title` and `icon` components. 
2. CSS change
  - Background color is now `#DDDDDD`. 
3. Explicitly list imported names
  - Components in this branch now only imports variables they need.
4.  More locales
  - More text, more locale edits. 
5. More icons
  - Right now only Github icon is added. 
  - If we require more icons, we might try using [fontawesome ](https://fontawesome.com/?from=io)instead of inserting images.


# Notes
`{this.props.children}` is react's  built in props.
See: https://reactjs.org/docs/composition-vs-inheritance.html#containment
Basically using this prop allows nesting components.
What I want to do here is 
```
<Card>
<OptionToggle></OptionToggle>
</Card>
```
But without `{this.props.children}`, `<OptionToggle>` will not render.

Bartleby2718 commented on src/constants.ts:
```diff
@@ -11,3 +11,13 @@ export const LOAD_BUTTON_TEXT: string = chrome.i18n.getMessage('loadButton');
+export const OPTIONS_ABOUT_TEAM_TITLE: string = chrome.i18n.getMessage('optionsAboutCardContentTeamNameTitle');
+export const OPTIONS_ABOUT_TEAM_CONTENT: string = chrome.i18n.getMessage('optionsAboutCardContentTeamNameContent');
+export const OPTIONS_ABOUT_TEAM_LINK_TITLE: string = chrome.i18n.getMessage('optionsAboutCardContentTeamLinkTitle');
+export const OPTIONS_ABOUT_TEAM_LINK_CONTENT: string = chrome.i18n.getMessage('optionsAboutCardContentTeamLinkContent');
```
Locale does not matter for this `string`(link). I think this should be defined here(`constants.ts`), not taken from `messages.json`.

jeongm-in replied:
huh good point. Will move them to `constants.ts`

Bartleby2718 commented on public/_locales/ko/messages.json:
```diff
@@ -22,5 +22,29 @@
+    "optionsAboutCardContentTeamLinkTitle": {
+        "message": "홈페이지"
+    },
+    "optionsAboutCardContentTeamLink": {
```
Typo, although I think this should be deleted in the first place.

Bartleby2718 commented on src/Option.tsx:
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import OptionPage from './components/OptionPage';
+// import { saveClicked, loadClicked } from './utils'
```
I can see that you've copied and pasted the entire file, but...

jeongm-in replied:
haha yeah

Bartleby2718 commented on src/components/Card.tsx:
```diff
@@ -0,0 +1,38 @@
+            <div style={this.style}>
+                <div style={this.textStyle}>{this.title}</div>
+                <hr></hr>
+                {this.props.children}
```
Where did `children` come from?

jeongm-in replied:
`{this.props.children}` is react's  built in props.
See: https://reactjs.org/docs/composition-vs-inheritance.html#containment
Basically using this prop allows nesting components.
What I want to do here is 
```
<Card>
<OptionToggle></OptionToggle>
</Card>
```
But without `{this.props.children}`, `<OptionToggle>` will not render.

Bartleby2718 replied:
This is the kind of thing I want you to write in pull request. The reviewer only sees the code and your pull request, not what you googled or learned.

Bartleby2718 commented on src/components/Logo.tsx:
```diff
@@ -0,0 +1,21 @@
+    image: string
+}
+
+class Logo extends React.Component<Props> {
```
I think if you were to be truly modular, this should be `Image`, not `Logo`. All it contains is a URL anyway.

Bartleby2718 commented on src/components/OptionToggle.tsx:
```diff
@@ -0,0 +1,67 @@
+                <span>{this.optionLabel}</span>
+                <Toggle
+                    checked={this.state.setting}
+                    icons={false}
```
What is this?

jeongm-in replied:
what is this in what is this?

jeongm-in replied:
https://github.com/aaronshaf/react-toggle#props

jeongm-in replied:
`icons`
>  If false, no icons are displayed. You may also pass custom icon components in icons={{{checked: <CheckedIcon />, unchecked: <UncheckedIcon />}}

Bartleby2718 commented on src/components/TitleField.tsx:
```diff
@@ -0,0 +1,27 @@
+    textContent: string,
+}
+
+class TextField extends React.Component<TextProps> {
```
Why is the file name `TitleField.tsx`?

Bartleby2718 commented on src/components/TitleField.tsx:
```diff
@@ -0,0 +1,27 @@
+}
+
+class TextField extends React.Component<TextProps> {
+    textTitle: string;
```
The attributes could have been more concise, since the class name already has `title` in it.

jeongm-in replied:
you mean `textTitle` to `title`? but this verbose name was intentional so as not to interfere with default HTML attributes.

Bartleby2718 replied:
Then why did you use `title` for `src/components/Card.tsx`?

jeongm-in replied:
Well one thing or another should be changed then

jeongm-in replied:
`title` :arrow_right:  `textTitle`

Bartleby2718 commented on webpack.config.js:
```diff
@@ -18,7 +19,14 @@ module.exports = {
+      {
+        // Preprocess 3rd party .css files located in node_modules
+        test: /\.css$/,
+        include: /node_modules/,
```
https://github.com/jeongm-in/webmark/issues/75#issuecomment-509733761
and the empty line below

Bartleby2718 commented on public/_locales/en/messages.json:
```diff
@@ -22,5 +22,29 @@
+        "message": "strgzr"
+    },
+    "optionsAboutCardContentTeamLinkTitle": {
+        "message": "Homepage"
```
Is this a good word? I can't think of a better word, but I feel like this is often used in Konglish.

Bartleby2718 commented on public/_locales/en/messages.json:
```diff
@@ -22,5 +22,29 @@
         "message": "Page was saved successfully."
+    },
+    "optionsGeneralCardTitle": {
+        "message": "General"
```
Shouldn't it be just `Settings`? We don't even have `Advanced (Settings)`. Also, you wrote `일반 설정` for the Korean translation.

jeongm-in replied:
valid. Hard to come up with good names (variable / values / keys)

Bartleby2718 replied:
agreed

Bartleby2718 commented on public/_locales/ko/messages.json:
```diff
@@ -22,5 +22,29 @@
+        "message": "일반 설정"
+    },
+    "optionsIsLoadInNewTab": {
+        "message": "저장된 페이지를 현재 창에서 열어요."
```
Shouldn't it be `탭`?

jeongm-in replied:
what is a good translation for tab?

Bartleby2718 replied:
Just `탭` would be fine, I think.

Bartleby2718 commented on public/_locales/ko/messages.json:
```diff
@@ -22,5 +22,29 @@
+    "optionsAboutCardTitle": {
+        "message": "소개"
+    },
+    "optionsAboutCardContentVersionText": {
```
typo

Bartleby2718 commented on public/_locales/ko/messages.json:
```diff
@@ -22,5 +22,29 @@
+    "optionsAboutCardContentVersionText": {
+        "message": "현재 버전"
+    },
+    "optionsAboutCardContentTeamNameText": {
```
typo

Bartleby2718 reviewed (changes requested):
I didn't really understand `componentDidMount`, but anyway well done.

jeongm-in committed:
Remove unnecessary rule from webpack.config.js close #75

jeongm-in committed:
Install Fontawesome libary for icons close #85

jeongm-in committed:
Refactor and edit typos for options #86

jeongm-in committed:
Rename and add more components for modularity close #86

jeongm-in commented:
- Moved link and team name to `constants.ts`
  - Added link to the Google Groups as well
- Achieve more modularity by splitting existing components into `container`, `image`, `text`
- Add [Fontawesome ](https://github.com/FortAwesome/react-fontawesome) libary to utilize icons
  - Sadly Fontawesome does not support passing `props` value to fetch component.
  - That is, in `GithubIcons.tsx` you can see the following block
     ```
     import { IconLookup, IconDefinition, findIconDefinition} from '@fortawesome/fontawesome-svg-core'
      ...

       render() {
        const githubLookup: IconLookup = { prefix: 'fab', iconName: 'github' }
        const githubIconDefinition: IconDefinition = findIconDefinition(githubLookup)
        return (
            <FontAwesomeIcon icon={githubIconDefinition} style={this.iconStyle} />
        );
    }
    ```
  - This is grabbed from the fontawesome-react documentations. Here you might think we could pass something like `this.props.iconTarget` into `iconName` field of `githubLookup: IconLookup`, so that one component could be used for multiple icons with different prop values.
  - The problem is, you will run into following error.
    ```    
    **(property) IconLookup.iconName: IconName**
    Type 'string' is not assignable to type 'IconName'.ts(2322)
    index.d.ts(6, 3): The expected type comes from property 'iconName' which is declared here on type 'IconLookup'
    ```
  - unless we provide an explicit, string value into `githubLookup: IconLookup`, this error will persist. 
  - Also, documentation mentions using `import { fab} from '@fortawesome/free-brands-svg-icons';` and `library.add(fab);` to load every "**f**amous-**b**rand`"icons. This caused `option.js` becoming too large:
    ```
    WARNING in webpack performance recommendations:
    You can limit the size of your bundles by using import() or require.ensure to lazy load some parts of your application. For more info visit https://webpack.js.org/guides/code-splitting/
        WARNING in asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
    This can impact web performance.
    Assets:
      option.js (603 KiB)
    
    WARNING in entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.
    Entrypoints:
      option (603 KiB)
          option.js
    ```
  - so I only imported `faGithub`.

- Edit css
  - In `Options.tsx`, it now has following line: `document.body.style.backgroundColor = "#DDD";`.
  - the style addition for the background color was originally placed in `OptionPage.tsx`. However, in a bigger monitor I found out this background color is only applied to the `<div>` of the `OptionPage` component, leaving about 1/4 of the screen blank white. 
  - turns out `body` of the html should have the `background-color` css property, so above line was added.
- Edited Webpack
- Edited manifest.json
  - this time also tested on ko locale

jeongm-in committed:
Remove unused dependency @fortawesome/free-solid-svg-icons #85

jeongm-in commented:
- Removed unused dependency `@fortawesome/free-solid-svg-icons`

Bartleby2718 commented on webpack.config.js:
```diff
@@ -4,20 +4,21 @@ module.exports = {
   mode: 'production',
   entry: {
     'background': './src/background.ts',
+    'option': './src/Option.tsx'
```
I just realized that Google uses `options.html`/`options.js` and we use `options.html`/`option.js`. Can we change this to `'options': './src/Option.tsx'`? If so, the `<script>` tag in `options.html` should be changed accordingly.

jeongm-in replied:
Y

Bartleby2718 replied:
What did I expect...

jeongm-in replied:
By the way Y meant Yes

Bartleby2718 replied:
🆗

Bartleby2718 commented on src/constants.ts:
```diff
@@ -4,10 +4,24 @@
 
+// locale independent values
+export const OPTIONS_ABOUT_TEAM_LINK_GITHUB: string = 'https://github.com/jeongm-in/webmark/';
+export const OPTIONS_ABOUT_TEAM_LINK_GOOGLE_GROUP: string = 'https://groups.google.com/forum/#!forum/strgzrs';
```
I'm sorry I wasn't specific. I wanted to add `<a href="mailto:strgzrs@googlegroups.com"`
https://en.wikipedia.org/wiki/Mailto
BTW anyone outside the group cannot access the group.

jeongm-in replied:
Okay that's what I thought. By the way some groups are open to the public. Would the public group be something we could use? Or is Google groups any relevant at all in the first place?

Bartleby2718 replied:
I don't think Google Groups has ever been and ever will be popular or even known to the public. I just wanted a way for the users to contact us. We can just use it as a mailing list.

Bartleby2718 commented on src/components/OptionPage.tsx:
```diff
@@ -0,0 +1,22 @@
+import * as React from "react";
+import ImageItem from "./ImageItem";
```
It just occurred to me that we might be reinventing the wheel, and sure enough, `react-bootstrap`(which is already included in the package) has an [`Image` component](https://react-bootstrap.github.io/components/images/). Take a look and see if this would be better. (If you do decide to use `react-bootstrap`, I'm so sorry for wasting your time. I kept googling without bootstrap and got nothing.)

Bartleby2718 replied:
They even have `Card` and many other components.

Bartleby2718 replied:
BTW I used `Button` and `ButtonGroup` in the popup (`index.html`).

jeongm-in replied:
Okay, so that's why components were able to be nested without `{this.props.children}`

jeongm-in replied:
I am going to leave `Card` as is, but replace image and maybe text with bootstrap components

Bartleby2718 commented on src/components/OptionToggle.tsx:
```diff
@@ -0,0 +1,67 @@
+    [key: string]: any
+}
+
+class OptionToggle extends React.Component<OptionToggleProps, OptionToggleState> {
```
If the parameter `OptionToggleState` is not used, why is it separated from `OptionToggleProps`?

jeongm-in replied:
State is used? It's used to store the value of the storage value. No?

Bartleby2718 replied:
I never said `state` is not used; I said `OptionToggleState` is not used.

jeongm-in replied:
According to typescript documentations I referred to without `OptionToggleState` state would not work. `Property 'setting' does not exist on type 'Readonly<{}>'.ts(2339)`

Bartleby2718 replied:
OK, so `interface Component` has two optional parameters `P`(rops) and `S`(tates), and we need an object `S`(with `state`s we need as its keys) as the second parameter in order to use states.
https://stackoverflow.com/a/47562985/4370146

jeongm-in replied:
Yeah this got me first time because in plain React.js I didn't had to provide parameters to Component to use states. Not that I am 100% sure my React coding style was up to the standards though.

Bartleby2718 commented on src/components/TextItem.tsx:
```diff
@@ -0,0 +1,21 @@
+    textTitle: string,
+}
+
+class TextItem extends React.Component<TextProps> {
```
I don't really get why we need this class. It doesn't even have any styling. The only thing special about this class is that it uses [`<label>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label), but <label> is supposed to be used with `<input>`, and we're literally just writing some strings. Isn't this an unnecessary complication?

jeongm-in replied:
Yeah sure. It used to be a bigger component with multiple tags in it, and while breaking it down this became nothing more than a standalone label tag.

Bartleby2718 replied:
So why do we need this? It's like creating an object that only contains a string.

jeongm-in replied:
I am going to remove this and use plain html tags.

Bartleby2718 replied:
👌

Bartleby2718 commented on src/components/OptionEntry.tsx:
```diff
@@ -0,0 +1,17 @@
+
+    render() {
+        return (
+            <div className="d-flex flex-row justify-content-between">
```
All of a sudden, I think this `className` should be a class property because it is inherent to this class. Like this:
```
    className = 'd-flex flex-row justify-content-between';
    render() {
        return (
            <div className={this.className}>
                {this.props.children}
            </div>
        );
    }
```
Although this is basically no more than using one more variable...

Bartleby2718 replied:
If you do change it, the same should be done for other classes, too.

jeongm-in replied:
Agreed

Bartleby2718 commented on src/components/OptionPage.tsx:
```diff
@@ -0,0 +1,22 @@
+    render() {
+        return (
+            <div className="d-flex flex-column align-items-center justify-content-center">
+                <ImageItem image="./images/promo.png"></ImageItem>
```
How about we imitate the [documentation](https://facebook.github.io/create-react-app/docs/adding-images-fonts-and-files) and do `import promo from './images/promo.png';`?

Bartleby2718 reviewed (changes requested):
1. So why do need Font Awesome in the first place? Can't we just download GitHub mark from https://github.com/logos? 
2. My apologies, I noticed the white space on the bottom when the background was emerald, but I forgot to point that out.
3. Sorry for being a Monday-morning quarterback. I try to comment comprehensively the first time, but the more you look at it, the more ideas pop up. 
4. Is `"files.insertFinalNewline"` set to `true` in your VS Code `settings.json`? I noticed the absence of final newline characters in many of your files in this pull request.

jeongm-in commented:
>     1. So why do need Font Awesome in the first place? Can't we just download GitHub mark from https://github.com/logos?
> 
- fontswesome provides high resolution icon options, and there is no need to worry about image being blurry on different screens or scaling the image. Icons will be treated like font which make things a lot easier.  Also the rational behind using fontawesome was to make it easier later on when we add more icons in there (for example, Facebook, Instagram, etc although unlikely) 

>     2. My apologies, I noticed the white space on the bottom when the background was emerald, but I forgot to point that out.
> 
- it is gone now, right? My laptop screen was small enough for the div to fill the whole screen, but working on a different screen revealed this. 
>     3. Sorry for being a Monday-morning quarterback. I try to comment comprehensively the first time, but the more you look at it, the more ideas pop up.
> 
>     4. Is `"files.insertFinalNewline"` set to `true` in your VS Code `settings.json`? I noticed the absence of final newline characters in many of your files in this pull request.
- I think new setup doesn't have json fully configured. I will add that and reformat files

Bartleby2718 commented on src/components/GithubIcon.tsx:
```diff
@@ -0,0 +1,34 @@
+import * as React from "react";
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { library } from '@fortawesome/fontawesome-svg-core';
```
This can be grouped with the other `import` from `@fortawesome/fontawesome-svg-core`.

Bartleby2718 replied:
Never mind . Just see https://github.com/jeongm-in/webmark/issues/85#issuecomment-510087478

jeongm-in committed:
Rename webpack entry for options page to match Google extension guideline #86

jeongm-in committed:
Remove unnecessary components #86

jeongm-in committed:
Update link for the Google Groups to email address #76

jeongm-in committed:
Update prop name for Card component #86

jeongm-in committed:
Replace constants with variables and edit minor formatting issues #86

jeongm-in committed:
Remove unnecessary library and import statements #85

Bartleby2718 commented on src/components/GithubIcon.tsx:
```diff
@@ -1,14 +1,6 @@
-} from '@fortawesome/fontawesome-svg-core'
-
-library.add(faGithub);
+import { faGithub, faGit } from '@fortawesome/free-brands-svg-icons';
```
> 'faGit' is declared but its value is never read.ts(6133)

jeongm-in replied:
when was that added? Will be removed.

Bartleby2718 commented on src/components/OptionPage.tsx:
```diff
@@ -0,0 +1,26 @@
+import Image from 'react-bootstrap/Image';
+import CardGeneralSettings from "./CardGeneralSettings";
+import CardAbout from "./CardAbout";
+// import * as promo from "../images/promo.png";
```
Let's just delete this.

jeongm-in replied:
:+1:

Bartleby2718 reviewed (changes requested):
https://github.com/jeongm-in/webmark/pull/77#discussion_r301943663
Regarding `d-flex flex-column align-items-center justify-content-center`, I was talking about assigning it to a class property (declaring outside the `render()`), but it's fine. Still, I think you should get into the habit of using `const`s when reassignment is not expected. Just make the two minor changes, and I'll approve the merge commit.

jeongm-in committed:
Remove unused import #86

jeongm-in committed:
Use class property instead of variables for className attribute #86
